### ISSUE TITLE
No gradient propagation to generator for R2 penalty?

### ISSUE THREAD

**Issue by soshisnet** [brownvc/R3GAN#12]
Thank you for the code release! It's a wonderful paper, and I learned a lot by reading through it.

In the code [here](https://github.com/brownvc/R3GAN/blob/da2a8d8009098b5287914b4615fc35198504f75c/R3GAN/Trainer.py#L30-L31),  why do you use detach() right after running the generator?  This way, the generator parameters won't be update from R2 calculation? But Eq.3 in the paper suggests the parameters of both disc. and gen. have to be updated...
Any clarification would be appreciated! Thanks in advance!

```

def AccumulateDiscriminatorGradients(self, Noise, RealSamples, Conditions, Gamma, Scale=1, Preprocessor=lambda x: x):
        RealSamples = RealSamples.detach().requires_grad_(True)
        FakeSamples = self.Generator(Noise, Conditions).detach().requires_grad_(True)

```

**Comment by nick-bvc** [comment]
Both R1 and R2 are applied only to the discriminator. Eq.3 simply says that R2 _depends on_ the current generator because it regularizes the discriminator response on the current generative distribution.
